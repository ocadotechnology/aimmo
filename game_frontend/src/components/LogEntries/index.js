import styled from 'styled-components'
import React, { Component } from 'react'
import PropTypes from 'prop-types'

export const LogEntry = styled.li`
  list-style: none;
  padding: ${props => props.theme.spacing.unit}px;
`

export const StyledLogEntries = styled.ul`
  list-style: none;
  margin: 0;
  padding: 0;
`

export default class LogEntries extends Component {
  generateLogEntries () {
    const lastLogEntry = this.props.logs[this.props.logs.length - 1]
    const logEntries = this.props.logs.map(logEntry =>
      <LogEntry
        key={logEntry.timestamp}
        innerRef={this.setLastLogEntryRef(logEntry, lastLogEntry)}>
        {logEntry.log}
      </LogEntry>
    )
    return logEntries
  }

  setLastLogEntryRef (logEntry, lastLogEntry) {
    return node => {
      if (logEntry.timestamp === lastLogEntry.timestamp) {
        this.props.lastLogRef(node)
      }
    }
  }

  render () {
    return (
      <StyledLogEntries>
        {this.generateLogEntries()}
      </StyledLogEntries>
    )
  }
}

LogEntries.propTypes = {
  logs: PropTypes.arrayOf(
    PropTypes.shape({
      timestamp: PropTypes.int,
      log: PropTypes.string
    })),
  lastLogRef: PropTypes.func
}
