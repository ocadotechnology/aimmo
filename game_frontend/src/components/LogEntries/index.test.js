/* eslint-env jest */
import React from 'react'
import LogEntries, { StyledLogEntries, LogEntry } from 'components/LogEntries'
import createShallowWithTheme from 'testHelpers/createShallow'

describe('<LogEntries />', () => {
  it('renders correctly with logs', () => {
    const tree = createShallowWithTheme(
      <LogEntries logs={[{ timestamp: '1', log: 'hello' }, { timestamp: 2, log: 'bye' }]} />,
      'dark'
    )
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly without logs', () => {
    const tree = createShallowWithTheme(<LogEntries logs={[]} />, 'dark')
    expect(tree).toMatchSnapshot()
  })
})

describe('<StyledLogEntries />', () => {
  it('renders correctly', () => {
    const tree = createShallowWithTheme(
      <StyledLogEntries logs={[{ timestamp: '1', log: 'hello' }, { timestamp: 2, log: 'bye' }]} />,
      'dark'
    )
    expect(tree).toMatchSnapshot()
  })
})

describe('<LogEntry />', () => {
  it('renders correctly', () => {
    const tree = createShallowWithTheme(<LogEntry>hello</LogEntry>, 'dark')
    expect(tree).toMatchSnapshot()
  })
})
