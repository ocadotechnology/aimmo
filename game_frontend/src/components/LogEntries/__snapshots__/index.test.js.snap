// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LogEntries /> renders correctly with logs 1`] = `
<styled.ul>
  <styled.li
    innerRef={[Function]}
    key="1"
  >
    hello
  </styled.li>
  <styled.li
    innerRef={[Function]}
    key="2"
  >
    bye
  </styled.li>
</styled.ul>
`;

exports[`<LogEntries /> renders correctly without logs 1`] = `<styled.ul />`;

exports[`<LogEntry /> renders correctly 1`] = `
.c0 {
  list-style: none;
  padding: 8px;
}

<li
  className="c0"
>
  hello
</li>
`;

exports[`<StyledLogEntries /> renders correctly 1`] = `
.c0 {
  list-style: none;
  margin: 0;
  padding: 0;
}

<ul
  className="c0"
/>
`;
